src/fb/mod.rs

```diff
@@ -0,0 +1,367 @@
+
+use crate::{
+    prelude::*,
+    utils::{pad_to_4bytes, parser, writer},
+    RtcpPacket, RtcpPacketParser, RtcpParseError, RtcpWriteError,
+};
+
+pub mod nack;
+pub mod pli;
+pub mod rpsi;
+pub mod sli;
+
+#[derive(Debug, Default, PartialEq, Eq)]
+pub struct FciFeedbackPacketType {
+    transport: bool,
+    payload: bool,
+}
+
+impl FciFeedbackPacketType {
+    pub const NONE: Self = Self {
+        transport: false,
+        payload: false,
+    };
+    pub const TRANSPORT: Self = Self {
+        transport: true,
+        payload: false,
+    };
+    pub const PAYLOAD: Self = Self {
+        transport: false,
+        payload: true,
+    };
+}
+
+impl std::ops::BitOr<FciFeedbackPacketType> for FciFeedbackPacketType {
+    type Output = Self;
+    fn bitor(self, rhs: Self) -> Self::Output {
+        Self {
+            transport: self.transport | rhs.transport,
+            payload: self.payload | rhs.payload,
+        }
+    }
+}
+
+impl std::ops::BitAnd<FciFeedbackPacketType> for FciFeedbackPacketType {
+    type Output = Self;
+    fn bitand(self, rhs: Self) -> Self::Output {
+        Self {
+            transport: self.transport & rhs.transport,
+            payload: self.payload & rhs.payload,
+        }
+    }
+}
+
+/// A parsed (Transport) Feedback packet as specified in RFC 4585.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct TransportFeedback<'a> {
+    data: &'a [u8],
+}
+
+impl<'a> RtcpPacket for TransportFeedback<'a> {
+    const MIN_PACKET_LEN: usize = 12;
+    const PACKET_TYPE: u8 = 205;
+}
+
+impl<'a> RtcpPacketParser<'a> for TransportFeedback<'a> {
+    fn parse(data: &'a [u8]) -> Result<Self, RtcpParseError> {
+        parser::check_packet::<Self>(data)?;
+        Ok(Self { data })
+    }
+
+    #[inline(always)]
+    fn header_data(&self) -> [u8; 4] {
+        self.data[..4].try_into().unwrap()
+    }
+}
+
+impl<'a> TransportFeedback<'a> {
+    pub fn builder() -> TransportFeedbackBuilder {
+        TransportFeedbackBuilder::default()
+    }
+
+    pub fn padding(&self) -> Option<u8> {
+        parser::parse_padding(self.data)
+    }
+
+    pub fn sender_ssrc(&self) -> u32 {
+        parser::parse_ssrc(self.data)
+    }
+
+    pub fn media_ssrc(&self) -> u32 {
+        parser::parse_ssrc(&self.data[4..])
+    }
+
+    pub fn parse_fci<F: FciParser<'a>>(&self) -> Result<F, RtcpParseError> {
+        if F::PACKET_TYPE & FciFeedbackPacketType::TRANSPORT == FciFeedbackPacketType::NONE {
+            return Err(RtcpParseError::WrongImplementation);
+        }
+        if parser::parse_count(self.data) != F::FCI_FORMAT {
+            return Err(RtcpParseError::WrongImplementation);
+        }
+        F::parse(&self.data[12..])
+    }
+}
+
+/// TransportFeedback packet builder
+#[derive(Debug, Default)]
+#[must_use = "The builder must be built to be used"]
+pub struct TransportFeedbackBuilder {
+    padding: u8,
+    sender_ssrc: u32,
+    media_ssrc: u32,
+    fci: Option<Box<dyn FciBuilder<'static>>>,
```

Comment:
We could use a `Cow` and add `to_owned()` and `fci_owned()` so we only heap-allocate if `TransportFeedbackBuilder` is moved to a different stack frame than where `FciBuilder` is instantiated.

Alternatively, if it's not meant to be part of another `struct`, `TransportFeedbackBuilder` could use a type parameter `T: FciBuilder<'a>`, so we don't rely on dynamic dispatch.

Reply:
My initial implementation used generics, however the `Packet` enum will not play well with that as it would require all variants as separate enum values.  Hence why I went with dynamic dispatch instead.

Reply:
I cannot easily seem to be able to make `FciBuilder` as `Cow` (with a `Sized` constraint) while also keeping `FciBuilder` object safe.

Reply:
> the Packet enum will not play well with that as it would require all variants as separate enum values

Yes, indeed.

> I cannot easily seem to be able to make FciBuilder as Cow (with a Sized constraint) while also keeping FciBuilder object safe.

I played with it a bit and can see the problems now. I think the only solution would be to build our own `FciBuilderCow` that would not rely on `ToOwned`. We would add an `into_owned(&self) -> Box<dyn FciBuilder<'static>>` method to the trait `FciBuilder<'a>` which would be used by `FciBuilderCow`. The new `into_owned()` could be auto-implemented for types which can derive `Clone` (such as `PliBuilder`) and we would explicitly implement it for others (such as for `RpsiBuilder`). I think that should tick all the boxes. I can experiment with this in a separate PR if you want.